### PR TITLE
Establish a nightly build against Theia latest main branch #163

### PR DESCRIPTION
**Pull request by jfaltermeier**
#### What it does
This is the first part of solving issue #163
What is missing in this PR is executing the tests on all platforms (currently only executed for Windows). We only deal with compilation and building the packages for now. 

After this is merged the master branch has a script that may be used to update all theia versions to `next`. This may be executed running `yarn update:next` in the repository root.

We will set up an integration job that will build Blueprint against Theia@next so that we may identify breaking changes early. Currently this is tested here: https://ci.eclipse.org/theia/job/theia-next/job/jf%252Fnext/

##### Proposed Build Process on Jenkins

* Check out the master branch containing all of the latest changes.
* Merge the next branch into the checked out branch. The next branch will only contain fixes that are required to build against the `next` version. The is set up here: https://github.com/eclipse-theia/theia-blueprint/tree/next 
~~At the moment this is needed because a method we use is renamed with Theia 1.19.0.~~ (Since we've updated to 1.189 by now, the next branch does not yet have any changes, but this was tested for the 1.19 update on the branch)
* Run `yarn update:next` (may have been executed on `next` already but not necessarily)
* Build blueprint and run the tests
* In case of an error notify responsible persons via e-mail

This process should make sure that we always include all of the latest changes (by using the master branch as a starting point for the build) and only have to maintain fixes (on the `next` branch) in case of actual problems.

##### Proposed way of handling errors

* Analyse the issue and open a bug report
* Either fix the issue (open a PR and merge to `next` branch) or create some kind of hotfix for the `next` branch to get a green build.

#### How to test
* run `yarn update:next` and check resulting changes to verify the package.json update
* review the proposed build process on Jenkins

#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

**Comment by marcdumais-work** [comment]
@jfaltermeier I think you will need to add `xvnc` to the `theia-dev` docker image: 
https://github.com/eclipse-theia/theia-blueprint/blob/master/Dockerfile

Once the Dockerfile change is merged, the image will be rebuild/published the next night:
https://github.com/eclipse-theia/theia-blueprint/blob/master/.github/workflows/publish-builder-img.yml

**Comment by jfaltermeier** [comment]
> @jfaltermeier I think you will need to add `xvnc` to the `theia-dev` docker image: https://github.com/eclipse-theia/theia-blueprint/blob/master/Dockerfile
> 
> Once the Dockerfile change is merged, the image will be rebuild/published the next night: https://github.com/eclipse-theia/theia-blueprint/blob/master/.github/workflows/publish-builder-img.yml

Thanks, I've added it and published it as jfaltermeier/theia-blueprint-build for now until the build is stable. 
A different issue was, that the $HOME directory is not writable when executed on the Eclipse Infrastructure. So the electron app could not start as it tries to write some files. Setting $HOME to a writable locations seems to work as a workaround: https://ci.eclipse.org/theia/job/theia-next/job/jf%252Fnext/14/execution/node/85/log/

This has 1 passing test for Linux. However it seems to be very slow. 

Mac is stuck at `2021-11-12T14:58:39.071Z INFO devtools: no browser pages found, retrying...` at the moment. This might be an issue with write access as well. 

I will continue to investigate-

**Comment by marcdumais-work** [comment]
AFAIK this is the first time we run/test Blueprint on Jenkins, so it's not too surprising that there are obstacles. 

For your consideration: it would be a consideration to split this PR into two: adding the `next` nightly build and in a second PR, adding the test suite (and the related troubleshooting). I think the first PR could be ready to merge quickly.

**Comment by jfaltermeier** [comment]
I've updated the PR accordingly. For now tests are only executed on windows. 
The pipeline adds a `@daily` trigger. 
In the post part of the pipeline script I've only added myself to get an email in case of an unstable/failed build for now. I am not sure if we should add all interested users here or create a dedicated mailing-list instead.

File: next/Jenkinsfile
Comment (marcdumais-work) [inline next/Jenkinsfile:152]: accidental indent?

**Comment by marcdumais-work** [comment]
> I've only added myself to get an email in case of an unstable/failed build for now. I am not sure if we should add all interested users here or create a dedicated mailing-list instead.

Not sure what would work best. Maybe it's just me, but I get so many emails that it's not necessarily the best mechanism to get my attention. As an additional mechanism, maybe we can consider to add a badge in the root README to reflect the status of the latest nightly `next` build?

**Comment by marcdumais-work** [comment]
@jfaltermeier ATM the `next` branch in this repo is not "special" - it's not protected and can be force-reset by anyone with write access. For flexibility, maybe we keep it that way for now. But it's possible to have webmaster protect that branch and e.g. require a review before merge, if we ever want to. 

If we keep the `next` branch updated, whenever there's an issue with new @theia content, in theory, when comes time to switch to the newly released @theia, it should contain exactly what's needed so we can adopt that new release. i.e. it could be used as the base for the PR that will update Blueprint to the latest @theia.

**Review by marcdumais-work — APPROVED** [review]
LGTM, thanks @jfaltermeier

**Review by JonasHelming — APPROVED** [review]

**Comment by jfaltermeier** [comment]
Thanks, I've collected the open issues mentioned above as subtasks of the issue: https://github.com/eclipse-theia/theia-blueprint/issues/163

**Pull request merged**
